File: web/backend/nnm_web/security.py
"""Upload safety primitives: filename sanitization, path-traversal rejection,
extension allow-listing, size limits, and SHA-256 checksums.

These functions are deliberately pure and side-effect free so they can be unit
tested in isolation. Raw file *contents* are never logged or echoed anywhere.
"""

from __future__ import annotations

import hashlib
import re
import unicodedata
from pathlib import PurePosixPath

from .settings import (
    ALLOWED_EXTENSIONS,
    ANALYZABLE_EXTENSIONS,
    ARCHIVAL_ONLY_EXTENSIONS,
    CATALOG_ONLY_EXTENSIONS,
    FORMAT_LABELS,
)


class UploadValidationError(ValueError):
    """Raised when an upload fails a safety or format check."""


_SAFE_CHARS = re.compile(r"[^A-Za-z0-9._-]+")
_MULTI_DOT = re.compile(r"\.{2,}")


def sanitize_filename(raw_name: str) -> str:
    """Return a safe base filename with no directory components.

    - Strips any path (both POSIX and Windows separators), defeating traversal
      such as ``../../etc/passwd`` or ``..\\..\\secret``.
    - Normalizes Unicode and drops disallowed characters.
    - Collapses repeated dots so ``a..b`` cannot re-introduce traversal.
    - Rejects empty results.
    """
    if raw_name is None:
        raise UploadValidationError("Missing filename.")

    # Normalize Unicode to a canonical form, then drop non-ASCII.
    name = unicodedata.normalize("NFKD", raw_name)
    name = name.encode("ascii", "ignore").decode("ascii")

    # Take only the final path component, handling both separator styles.
    name = name.replace("\\", "/")
    name = PurePosixPath(name).name  # strips directories and leading slashes

    # Remove control chars and disallowed characters.
    name = name.replace("\x00", "")
    name = _SAFE_CHARS.sub("_", name).strip("._")
    name = _MULTI_DOT.sub(".", name)

    if not name or name in {".", ".."}:
        raise UploadValidationError("Filename is empty after sanitization.")
    # Cap the length to keep filesystem paths reasonable.
    if len(name) > 128:
        stem, _, ext = name.rpartition(".")
        name = (stem[:120] + "." + ext) if ext else name[:128]
    return name


def is_path_safe(candidate: str) -> bool:
    """Return True if ``candidate`` has no traversal or absolute components."""
    if not candidate:
        return False
    normalized = candidate.replace("\\", "/")
    if normalized.startswith("/"):
        return False
    parts = PurePosixPath(normalized).parts
    return ".." not in parts and not any(p.startswith("/") for p in parts)


def extension_of(filename: str) -> str:
    """Return the lowercased extension including the leading dot (or '')."""
    return PurePosixPath(filename).suffix.lower()


def classify_extension(ext: str) -> str:
    """Return an analysis-mode string for a (validated) extension."""
    ext = ext.lower()
    if ext in ANALYZABLE_EXTENSIONS:
        return "analyze"
    if ext in CATALOG_ONLY_EXTENSIONS:
        return "catalog-only"
    if ext in ARCHIVAL_ONLY_EXTENSIONS:
        return "archival-only"
    return "unsupported"


def format_label(ext: str) -> str:
    return FORMAT_LABELS.get(ext.lower(), "Unknown")


def validate_extension(filename: str) -> str:
    """Validate the file extension against the allow-list.

    Returns the normalized extension. Raises :class:`UploadValidationError`
    for anything outside the allow-list.
    """
    ext = extension_of(filename)
    if ext not in ALLOWED_EXTENSIONS:
        allowed = ", ".join(sorted(ALLOWED_EXTENSIONS))
        raise UploadValidationError(
            f"Unsupported file type '{ext or '(none)'}'. Allowed: {allowed}."
        )
    return ext


def validate_size(size_bytes: int, max_bytes: int) -> None:
    """Reject empty or oversized uploads."""
    if size_bytes <= 0:
        raise UploadValidationError("Uploaded file is empty.")
    if size_bytes > max_bytes:
        mb = max_bytes / (1024 * 1024)
        raise UploadValidationError(
            f"File exceeds the {mb:.0f} MB upload limit (deployment proxy caps requests at 10 MB)."
        )


def sha256_of_bytes(data: bytes) -> str:
    """Compute a hex SHA-256 digest of raw bytes without logging contents."""
    h = hashlib.sha256()
    h.update(data)
    return h.hexdigest()


def sha256_of_file(path) -> str:
    """Compute a hex SHA-256 digest of a file, streamed in chunks."""
    h = hashlib.sha256()
    with open(path, "rb") as fh:
        for chunk in iter(lambda: fh.read(1024 * 1024), b""):
            h.update(chunk)
    return h.hexdigest()

File: web/vercel/lib/security.ts
/**
 * Upload safety primitives (TypeScript port of security.py):
 * filename sanitization, path-traversal rejection, extension allow-listing,
 * size limits, and SHA-256 checksums.
 *
 * Pure and side-effect free so they can be unit tested in isolation. Raw file
 * *contents* are never logged, echoed, or embedded in errors.
 */

import { createHash } from 'node:crypto';
import {
  ALLOWED_EXTENSIONS,
  ANALYZABLE_EXTENSIONS,
  ARCHIVAL_ONLY_EXTENSIONS,
  CATALOG_ONLY_EXTENSIONS,
  FORMAT_LABELS,
} from './config.js';

export class UploadValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'UploadValidationError';
  }
}

const SAFE_CHARS = /[^A-Za-z0-9._-]+/g;
const MULTI_DOT = /\.{2,}/g;

/** Return the final path component regardless of separator style. */
function baseName(name: string): string {
  const normalized = name.replace(/\\/g, '/');
  const parts = normalized.split('/');
  return parts[parts.length - 1] ?? '';
}

/**
 * Return a safe base filename with no directory components.
 * - Strips any path (POSIX and Windows), defeating traversal.
 * - Normalizes Unicode (NFKD) and drops non-ASCII.
 * - Collapses repeated dots so "a..b" cannot re-introduce traversal.
 * - Rejects empty results.
 */
export function sanitizeFilename(rawName: string | null | undefined): string {
  if (rawName == null) {
    throw new UploadValidationError('Missing filename.');
  }
  // Normalize Unicode to a canonical form, then drop non-ASCII.
  let name = rawName.normalize('NFKD').replace(/[^\x00-\x7F]/g, '');
  // Take only the final path component.
  name = baseName(name);
  // Remove control chars and disallowed characters.
  name = name.replace(/\x00/g, '');
  name = name.replace(SAFE_CHARS, '_').replace(/^[._]+|[._]+$/g, '');
  name = name.replace(MULTI_DOT, '.');

  if (!name || name === '.' || name === '..') {
    throw new UploadValidationError('Filename is empty after sanitization.');
  }
  if (name.length > 128) {
    const dot = name.lastIndexOf('.');
    if (dot > 0) {
      const stem = name.slice(0, dot);
      const ext = name.slice(dot + 1);
      name = `${stem.slice(0, 120)}.${ext}`;
    } else {
      name = name.slice(0, 128);
    }
  }
  return name;
}

/** True if `candidate` has no traversal or absolute components. */
export function isPathSafe(candidate: string): boolean {
  if (!candidate) return false;
  const normalized = candidate.replace(/\\/g, '/');
  if (normalized.startsWith('/')) return false;
  const parts = normalized.split('/');
  return !parts.includes('..') && !parts.some((p) => p.startsWith('/'));
}

/** Return the lowercased extension including the leading dot (or ''). */
export function extensionOf(filename: string): string {
  const base = baseName(filename);
  const dot = base.lastIndexOf('.');
  return dot > 0 ? base.slice(dot).toLowerCase() : '';
}

export type AnalysisMode = 'analyze' | 'catalog-only' | 'archival-only' | 'unsupported';

export function classifyExtension(ext: string): AnalysisMode {
  const e = ext.toLowerCase();
  if (ANALYZABLE_EXTENSIONS.has(e)) return 'analyze';
  if (CATALOG_ONLY_EXTENSIONS.has(e)) return 'catalog-only';
  if (ARCHIVAL_ONLY_EXTENSIONS.has(e)) return 'archival-only';
  return 'unsupported';
}

export function formatLabel(ext: string): string {
  return FORMAT_LABELS[ext.toLowerCase()] ?? 'Unknown';
}

/** Validate the extension against the allow-list; return the normalized ext. */
export function validateExtension(filename: string): string {
  const ext = extensionOf(filename);
  if (!ALLOWED_EXTENSIONS.has(ext)) {
    const allowed = [...ALLOWED_EXTENSIONS].sort().join(', ');
    throw new UploadValidationError(
      `Unsupported file type '${ext || '(none)'}'. Allowed: ${allowed}.`,
    );
  }
  return ext;
}

/** Reject empty or oversized uploads. */
export function validateSize(sizeBytes: number, maxBytes: number): void {
  if (sizeBytes <= 0) {
    throw new UploadValidationError('Uploaded file is empty.');
  }
  if (sizeBytes > maxBytes) {
    const mb = maxBytes / (1024 * 1024);
    throw new UploadValidationError(
      `File exceeds the ${mb.toFixed(0)} MB upload limit ` +
        '(Vercel caps serverless request bodies at ~4.5 MB).',
    );
  }
}

/** Compute a hex SHA-256 digest of raw bytes without logging contents. */
export function sha256OfBytes(data: Uint8Array): string {
  return createHash('sha256').update(data).digest('hex');
}
